### PR TITLE
Teach the worklog how to correct keyring memberships

### PR DESCRIPTION
**Pull request by jbowes**
If an invite or machine creation fails partway through, we can end up in
a state where the user/machine is created, but are not given access to
all keyrings, and the secrets therein.

Teach the worklog how to identify and resolve this problem.

### CODE REVIEW

**Comment by ianlivingstone** [comment]
Also needed for re-encoding users into keyrings in teh case of account recovery!

File: daemon/logic/worklog.go [hunk 1]
```diff
@@ -100,13 +107,30 @@ func (w *Worklog) Resolve(ctx context.Context, n *observer.Notifier,
 		return nil, nil
 	}
 
-	return w.handlers[item.Type()].resolve(ctx, n, orgID, item)
+	handler := w.handlers[item.Type()]
+	result, err := handler.resolve(ctx, n, orgID, item)
+
+	if err != nil {
+		result = &apitypes.WorklogResult{
```
Comment (ianlivingstone) [inline daemon/logic/worklog.go]: Why return a special result for an error instead of a standard error?

Comment (ianlivingstone) [inline daemon/logic/worklog.go]: Ah I see, it was here before, you jsut cleaned it up a bit. Still a question though! Perhaps add a comment?

File: daemon/logic/worklog.go [hunk 2]
```diff
@@ -242,17 +270,257 @@ func (h *inviteApproveHandler) resolve(ctx context.Context, n *observer.Notifier
...
 	if err != nil {
-		return &apitypes.WorklogResult{
-			ID:      item.ID,
-			State:   apitypes.ErrorWorklogResult,
-			Message: "Error approving invite: " + err.Error(),
-		}, nil
+		return nil, err
+	}
+
+	// Find all of the credential graphs in this org.
```
Comment (ianlivingstone) [inline daemon/logic/worklog.go]: You should be able to just `GET /keyrings` ?

Comment (ianlivingstone) [inline daemon/logic/worklog.go]: https://github.com/manifoldco/registry/blob/master/schema/request/keyrings/list/response.json

Returns everything *but* the credentials -- which we don't need!

Comment (ianlivingstone) [inline daemon/logic/worklog.go]: Thoughts? Perhaps you wanted to standardize on the the credentialgraph endpoint for caching?

File: daemon/logic/worklog.go [hunk 3]
```diff
@@ -242,17 +270,257 @@ func (h *inviteApproveHandler) resolve(ctx context.Context, n *observer.Notifier
...
+	//
+	// The paths map takes care of eliminating multiple versions of the keyring;
+	// the subsequent List call will return all versions.
+	cgs := newCredentialGraphSet()
+	orgName := org.Body.(*primitive.Org).Name
+	paths := make(map[string]*pathexp.PathExp)
+	for _, project := range projects {
+		projName := project.Body.(*primitive.Project).Name
+		graphs, err := h.engine.client.CredentialGraph.Search(ctx,
+			"/"+orgName+"/"+projName+"/*/*/*/*", h.engine.session.AuthID())
```
Comment (ianlivingstone) [inline daemon/logic/worklog.go]: ugh - i really wan to fix this tech debt ;(

File: daemon/logic/worklog.go [hunk 4]
```diff
@@ -242,17 +270,257 @@ func (h *inviteApproveHandler) resolve(ctx context.Context, n *observer.Notifier
...
+	// members of a keyring based on ACLs.
+	members, err := getKeyringMembers(ctx, h.engine.client, org.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	missing := make(map[string]apitypes.WorklogItem)
+	for _, graph := range graphs {
+		for _, member := range members {
+			m, _, err := graph.FindMember(&member)
```
Comment (ianlivingstone) [inline daemon/logic/worklog.go]: does this handle revoked keys?

Comment (ianlivingstone) [inline daemon/logic/worklog.go]: i.e. it says a user isn't a member if the included keys are revoked?

Comment (ianlivingstone) [inline daemon/logic/worklog.go]: (If you have a valid keypair but the currently encoded key is revoked, we should encode your new keys)

File: daemon/logic/worklog.go (hunk 1 above)
Comment (jbowes) [inline daemon/logic/worklog.go]: ya can do. it's just to differentiate between a fatal error in the daemon/registry vs an error in an individual worklog item; so if you're resolving many at once, you can keep going.

File: daemon/logic/worklog.go (hunk 3 above)
Comment (jbowes) [inline daemon/logic/worklog.go]: which part? ranging over projects?
The api endpoint says it should support `*` project name, but doesn't :(

File: daemon/logic/worklog.go (hunk 4 above)
Comment (jbowes) [inline daemon/logic/worklog.go]: this will take a lil' bit. i'll push up the other changes now.

File: daemon/logic/worklog.go (hunk 2 above)
Comment (ianlivingstone) [inline daemon/logic/worklog.go]: Right - forgot about the "activity" component.

File: daemon/logic/worklog.go (hunk 3 above)
Comment (ianlivingstone) [inline daemon/logic/worklog.go]: That exactly! It wouldn't be a ton of work to push this behaviour into the server, and we do this in many different places (torus ls).

File: daemon/logic/worklog.go (hunk 4 above)
Comment (ianlivingstone) [inline daemon/logic/worklog.go]: Ok - we can do it as a separate PR as well -- we don't have a good mechanism for revoking keypairs for testing either, perhaps we should implement a hidden `torus keypairs revoke` until we have a full fledged account recovery flow?

Comment (jbowes) [inline daemon/logic/worklog.go]: yeah i'll do it in another PR. lets get this party moving. https://github.com/manifoldco/torus-cli/issues/153
